Client can not specify command

rminnich opened the pull request:
also includes a fix.

rminnichcodeu committed:
cmds/cpu,cmds/cpud,packages: client does not set Cmd

Up until now, the cpu command has been allowed to specify
what cpud command is run.

This was fine on Plan 9, where the authentication was done
before cpud is run. On Unix, it is not fine, as cpud must
do substantial work (with privs) before setting uid/gid
for the user.

Remove client's ability to set the command. cpud server
will now invoke cpud remote with, always, cpud -remote
as the first two arguments.

Signed-off-by: Ronald G. Minnich <rminnich@google.com>

rminnichcodeu committed:
cpud: if len(args) is 0, assume os.Getenv("SHELL")

If SHELL is unset, cpud will exit with an error
message:

No arguments and $SHELL is not set

Signed-off-by: Ronald G. Minnich <rminnich@google.com>

rminnichcodeu committed:
decpu: remove cpudcmd option

no longer supported.

decpud will probably need to be fixed.

Signed-off-by: Ronald G. Minnich <rminnich@google.com>

Lencerf commented on server/server.go:
```diff
@@ -62,7 +62,7 @@ func errval(err error) error {
 	a := s.Command()
 	verbose("handler: cmd is %q", a)
-	cmd := command(a[0], a[1:]...)
+	cmd := command("cpud", append([]string{"-remote"}, a...)...)
```
How about using `os.Args[0]` instead of `cpud`?

rminnich replied:
Let me think on that. I am not sure. I can see reasons for both paths.

rminnich replied:
e.g., what if we started cpud as /init, but we really want to run with argv[0] = cpud. but, yeah, I'm tempted to like this idea, but ... hmm, hmm, hmmm, hmmmmmmmm, not sure.

brho replied:
from a debugging/hacking perspective, if i was trying something out and renamed my new version "cpud-new", but then it turns out it was calling the original cpud that was still in $PATH, i'd be gouging my eyeballs out.

ericvh replied:
yeah - I was just trying out the debugging use case like the other day.

Lencerf replied:
If we have `/init` linked to `/bin/cpud`, and start the cpud server as `/init`, when cpud starts the `-remote` subprocess, we will have `/init -remote <....>`, which will still work correctly.

rminnich replied:
I think I'd like to stick with cpud as the name to exec instead of argv[0]. But I can't quite tell from comments what brho and eric think. This allows playing with PATH and namespaces to get a new cpud from the server. But it's definitely not a clear decision.

Lencerf replied:
`server` is a package and both `cpud` and `decpud` use it. If someone uses `decpu` to talk to `decpud`, the `decpud` server will end up invoking `cpud` to launch a task.

rminnich replied:
yeah. That's another concern.

rminnichcodeu committed:
Have the cpud specify which cpud to run

The server package was blindly specifying "cpud", which
is clearly a mistake.

A proposed option was to use os.Args[0], which is concerning
as programs might change os.Args (for, e.g., security reasons)
and having a package blindly depend on that global is similary
worrisome.

Change server.New to accept a third argument, then path of the
cpud; and use a closure in the ssh handler so that argument
can be used by the ssh package.

Signed-off-by: Ronald G. Minnich <rminnich@google.com>

rminnich commented:
ok, look at this. I was not comfortable having a package blindly use os.Args

rminnichcodeu committed:
cmds/cpud: clean up wording error in comment

Signed-off-by: Ronald G. Minnich <rminnich@google.com>
